### PR TITLE
feat: make `to_compact` borrow

### LINKED ISSUE #9487: make Compact::encode accept &self
### Describe the feature

accepting `self` has only downsides

https://github.com/paradigmxyz/reth/blob/67f617646ce4b68a5096239238aa9209636ff500/crates/storage/codecs/src/lib.rs#L45-L47

for example, with #8742 we now operate on `Arc` values, making disk writes more expensive because we need to clone everything

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattsse**
closes #9487

needs more work on derive impls next

### CODE REVIEW

File: crates/storage/db/src/tables/mod.rs [hunk 1]
```diff
@@ -381,7 +383,7 @@ tables! {
     table HashedStorages<Key = B256, Value = StorageEntry, SubKey = B256>;
 
     /// Stores the current state's Merkle Patricia Tree.
-    table AccountsTrie<Key = StoredNibbles, Value = StoredBranchNode>;
+    table AccountsTrie<Key = StoredNibbles, Value = StoredBranchNodeStatic>;
```
Comment (mattsse) [inline crates/storage/db/src/tables/mod.rs:386]: this is weird, why do we need to change the branchnode type?

File: crates/trie/common/src/hash_builder/value.rs [hunk 2]
```diff
@@ -2,16 +2,17 @@ use alloy_primitives::B256;
 use alloy_trie::hash_builder::HashBuilderValue;
 use bytes::Buf;
 use reth_codecs::Compact;
+use std::borrow::Cow;
 
 /// A wrapper around `HashBuilderValue` that implements `Compact`.
-pub(crate) struct StoredHashBuilderValue(pub(crate) HashBuilderValue);
+pub(crate) struct StoredHashBuilderValue<'a>(pub(crate) Cow<'a, HashBuilderValue>);
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/trie/common/src/hash_builder/value.rs:8]: hmm, not super pretty but will do

**Review by mattsse — COMMENTED** [review]
can't review my own pr hehe

overall I think this is an improvement, but this really shows that the compact trait has a lot of shortcomings 
I'd like to see some settings documented on the proc macro
and have a few questions

and would be nice if the vec specialization would also work for slices

File: crates/storage/db-api/src/models/client_version.rs [hunk 3]
```diff
@@ -23,14 +23,14 @@ impl ClientVersion {
...
     where
         B: bytes::BufMut + AsMut<[u8]>,
     {
         let Self { version, git_sha, build_timestamp } = self;
-        version.into_bytes().to_compact(buf);
-        git_sha.into_bytes().to_compact(buf);
-        build_timestamp.into_bytes().to_compact(buf)
+        version.as_bytes().to_vec().to_compact(buf);
+        git_sha.as_bytes().to_vec().to_compact(buf);
+        build_timestamp.as_bytes().to_vec().to_compact(buf)
```
Comment (mattsse) [inline crates/storage/db-api/src/models/client_version.rs:33]: is the to_vec still required?

**Review by mattsse — COMMENTED** [review]
lgtm

**Review by joshieDo — APPROVED** [review]

**Review by Rjected — APPROVED** [review]
cool, this makes sense to me, let's make an issue for splitting the trait

**Pull request merged**